File: boson-java/src/main/java/io/zink/joson/impl/JosonValidate.java
package io.zink.joson.impl;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.module.SimpleModule;
import de.undercouch.bson4jackson.BsonFactory;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import io.zink.boson.bson.bsonImpl.BosonImpl;
import io.zink.boson.bson.bsonPath.Interpreter;
import io.zink.boson.bson.bsonPath.Program;
import io.zink.boson.bson.bsonPath.TinyLanguage;
import io.zink.joson.Joson;
import scala.Option;
import scala.util.parsing.combinator.Parsers;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.concurrent.CompletableFuture;
import java.util.function.Consumer;
import java.util.function.Function;

public class JosonValidate<T> implements Joson {

    private String expression;
    private Consumer<T> validateFunction;


    public JosonValidate(String expression, Consumer<T> validateFunction) {
        this.expression = expression;
        this.validateFunction = validateFunction;
    }


    private void callParse(BosonImpl boson, String expression){
        TinyLanguage parser = new TinyLanguage();
        try{
            Parsers.ParseResult pr = parser.parseAll(parser.program(), expression);
            if(pr.successful()){
                Interpreter interpreter = new Interpreter(boson, (Program) pr.get(), Option.empty(), Option.apply(validateFunction), Option.empty());
                interpreter.run();
            }else{
                throw new RuntimeException("Failure/Error parsing!");
            }
        }catch (RuntimeException e){
            throw new RuntimeException(e.getMessage());
        }
    }


    @Override
    public CompletableFuture<String> go(String jsonStr) {
        JsonObject a = new JsonObject(jsonStr);
        ObjectMapper mapper = new ObjectMapper(new BsonFactory());
        ByteArrayOutputStream os = new ByteArrayOutputStream();
        SimpleModule module = new SimpleModule();
        module.addSerializer(JsonObject.class, new Joson.JsonObjectSerializer());
        module.addSerializer(JsonArray.class, new Joson.JsonArraySerializer());
        mapper.registerModule(module);
        byte[] bsonByteEncoding;
        CompletableFuture<String> future;
        try {
            mapper.writeValue(os, a);
            bsonByteEncoding = os.toByteArray();
            os.flush();
            future =
                    CompletableFuture.supplyAsync(() -> {
                        Option<byte[]> opt = Option.apply(bsonByteEncoding);
                        Option e = Option.empty();
                        BosonImpl boson = new BosonImpl(opt, e,e);
                        callParse(boson, expression);
                        //validateFunction.accept((T)value);
                        return jsonStr;
                    });
        }catch (IOException e){
            System.out.println(e.getMessage());
            future = new CompletableFuture<>();
            future.complete(jsonStr);

        }
        return future;
    }


    @Override
    public Joson fuse(Joson joson) {
        return null;
    }
}

File: boson-scala/src/main/scala/io/zink/joson/impl/JosonValidate.scala
//package io.zink.joson.impl
//
//import java.io.ByteArrayOutputStream
//
//import com.fasterxml.jackson.databind.ObjectMapper
//import com.fasterxml.jackson.databind.module.SimpleModule
//import de.undercouch.bson4jackson.BsonFactory
//import io.vertx.core.json.{JsonArray, JsonObject}
//import io.zink.boson.bson.bsonImpl.BosonImpl
//import io.zink.boson.bson.bsonPath.{Interpreter, Program, TinyLanguage}
//import shapeless.Typeable
////import io.zink.boson.bson.bsonValue.{BsObject, BsValue}
//import io.zink.joson.Joson
//import io.zink.joson.Joson.{JsonArraySerializer, JsonObjectSerializer}
//
//import scala.concurrent.ExecutionContext.Implicits.global
//import scala.concurrent.Future
//
//
//class JosonValidate[T](expression: String, validateFunction: T => Unit) extends Joson {
//  private def callParse(boson: BosonImpl, expression: String): Unit = {
//    val parser = new TinyLanguage
//    try {
//      parser.parseAll(parser.program, expression) match {
//        case parser.Success(r, _) =>
//          //new Interpreter[T](boson, r.asInstanceOf[Program], fExt = Option(validateFunction)).run()
//        case parser.Error(msg, _) =>
//          throw new Exception(msg)
//          //BsObject.toBson(msg)
//        case parser.Failure(msg, _) =>
//          throw new Exception(msg)
//          //BsObject.toBson(msg)
//      }
//    } catch {
//      case e: RuntimeException =>
//        throw new Exception(e.getMessage)
//        //BsObject.toBson(e.getMessage)
//    }
//  }
//
//  /**
//    * Apply this Joson to the String that arrives and at some point in the future complete
//    * the future with the resulting String. In the case of an Extractor this will result in
//    * the immutable String being returned unmodified.
//    *
//    * @param the Json string.
//    * @return
//    */
//  override def go(jsonStr: String): Future[String] = {
//    val a: JsonObject = new JsonObject(jsonStr)
//
//    val mapper: ObjectMapper = new ObjectMapper(new BsonFactory())
//    val os = new ByteArrayOutputStream
//    val module = new SimpleModule
//    module.addSerializer(classOf[JsonObject],new JsonObjectSerializer)
//    module.addSerializer(classOf[JsonArray], new JsonArraySerializer)
//    mapper.registerModule(module)
//
//    mapper.writeValue(os, a)
//
//    val bsonByteEncoding: Array[Byte] = os.toByteArray
//    os.flush()
//
//    val future: Future[String] =
//      Future{
//        val boson:BosonImpl = new BosonImpl(byteArray = Option(bsonByteEncoding))
//        callParse(boson, expression)
////        match {
////          case (res: BsValue) =>
////            validateFunction(res.asInstanceOf[T])
////          case _ => throw new RuntimeException("JosonExtractor -> go() default case!!!")
////        }
//        jsonStr
//      }
//    future
//  }
//
//  /**
//    * Fuse one Joson to another. The joson that is 'this' should be executed first before the
//    * joson that is the parameter in teh case of update/read conflicts.
//    * the immutable String being returned unmodified.
//    *
//    * @param the Joson to fuse to.
//    * @return the fused Joson
//    */
//  override def fuse(joson: Joson): Joson = new JosonFuse(this, joson)
//}
